[Vending machine step4] 앱 생명주기와 객체 저장

1Consumption opened the pull request:
scene이 연결되기 직전인 
``` swift
 func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) 
```
메소드 호출시에 UserDefualt에서 data를 불러오게 했으며, 

scene이 inActive 되는 시점인
```swift
func sceneWillResignActive(_ scene: UIScene)
```
메소드 호출시에 UserDefault에 data를 저장하게 하였습니다.

이 과정에서 미처 발견하지 못하여 무척이나 애를 먹었던 부분이 있었는데, 바로 enum 타입 프로퍼티를 아카이빙 하는 문제였습니다. 이 문제는 swift enum은 objective C에서 호환이 되지 않기 때문임을 알았습니다. NSCoding 과 NSObject 프로토콜을 채택한 encoding, decoding이기 때문에 위와 같은 문제가 발생했다고 생각합니다. 따라서 이 부분을 해결하기 위해 enum 타입을 아카이빙하는 것이 아닌  rawValue를 아카이빙하여 후에 decode해줄 때 rawValue를 통해 init을 하여 프로퍼티를 초기화 해줬습니다.

1Consumption committed:
[기능 수정] UI 변경

1Consumption committed:
[기능 추가] 클래스 직렬화

인스턴스를 직렬화 해서 로컬 저장소에 저장하고 불러오기 위해
NSCoding, NSObject 프로토콜을 채용하여 encode, decode 해줌

1Consumption committed:
[기능 추가] 데이터 저장 시점 설정

scene이 inactive 되면 데이터를 저장하게 하였고
scene이 연결되기전에 데이터를 불러오게 설정

1Consumption committed:
[기능 수정] enum 타입 아카이빙 설정

swift enum은 objevtive c에서 호환이 안되기 때문에 아카이빙이 안되는 오류 발생
이를 해결하기 위해 enum타입을 아카이빙 하지 않고 rawValue를 아카이빙 하여 디코딩할 때 rawValue를 매핑해주는 방식으로 변경

1Consumption committed:
[기능 수정] didSet 적용

프로퍼티에 didSet을 적용하여 값이 변하면 Notification을 보내게 설정

1Consumption committed:
[기능 수정] ReadMe 수정

스크린샷 첨부
주요 로직 설명 추가

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -68,14 +103,14 @@ class VendingMachine{
     
     func insert(beverageNumber: Int) {
-        beverages.add(beverage: beverageList[beverageNumber])
+        beverages.add(beverage: VendingMachine.beverageList[beverageNumber])
```
이 부분에서 자판기 음료수 화면 배치가 바꿔야 한다면 얼마나 고쳐야 하나요? 
자판기 동작이 음료수 순서와 독립적으로 구현할 수는 없을까요?

1Consumption replied:
ViewController에서는 버튼의 인덱스만 알고 있고, Model에서는 버튼의 인덱스를 모릅니다. 그래서 이를 매핑하기 위해 BeveragesList라는 static 배열을 선언해줬고요. 그래서 음료 배열의 순서와 화면에 표시된 음료의 순서가 같습니다. 말씀하신 것처럼 독립적이지 못합니다. 화면 배치가 바뀐다면 배열의 순서도 같이 바꿔줘야 합니다. 
하지만 이 방법 외에는 떠오르지 않아서 부득이하게 이런 방식으로 구현하였습니다. 혹시 독립적으로 구현할 수 있는 힌트가 될만한 키워드를 알려주실 수 있으신가요?

godrm replied:
VendingMachine 객체가 알고 있어야 하는 것과 ViewController 객체가 알고 있어야 하는 것을 구분하는 것부터 시작해보죠. 
화면 위치는 ViewController가 알아야 하는 것이고, 자판기는 음료수만 알면 되지 않을까요?
그러면 화면 배열과 음료수를 매칭하는 것을 누가 갖고 있어야 하는지 생각해보세요. 
지금 당장 완벽한 방법을 찾기보다는 데이터와 소유권, 알고있어야 하는 범위를 고려해서 포함하면 됩니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Price.swift:
```diff
@@ -61,9 +86,3 @@ class Price {
         return price.money >= self.money
     }
 }
```
위에 add, minus 에서도 money를 사용하게 isCheaper() 와 다르네요. 
Int대신 동일하게 할 수 있지 않을까요?

1Consumption replied:
이 질문은 잘 이해하지 못하겠습니다. 혹시 조금만 더 설명해주실 수 있으실까요?

godrm replied:
money에 접근하는 방식, money 값을 처리하는 타입이 다르다는 의미입니다. 
Price 타입을 그대로 활용하면 좋을 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Price.swift:
```diff
@@ -8,13 +8,38 @@
+    }
+    
+    func encode(with coder: NSCoder) {
+        coder.encode(money, forKey: "\(Property.money)")
```
이 정도면 Property.moneyKey = "money" 문자열로 정의하는 것보다 훨씬 더 길어질 것 같은데요. 어떻게 생각해요?

1Consumption replied:
네, 저도 그렇게 생각합니다. enum을 쓰면서 굳이 CustomStringConvertible 프로토콜을 채용하면서 까지 String으로 표현을 해야하나 생각을 했습니다. 부자연스럽다고 생각이 되기 때문이죠.
따라서 Property 클래스를 생성하여 그 안에서 static 상수를 하나 선언하여 각 model 별로 가지고 있던 Property enum을 대체 해줬습니다. 덕분에 관리하기 한결 수월 해졌습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -8,13 +8,82 @@
+            self.name = name
+        } else {
+            self.name = ""
+        }
```
이런 표현을 대신 해주는 게 `??` 이죠. 참고하세요.

1Consumption replied:
덕분에 코드가 한결 깔끔해졌습니다!

godrm commented on VendingMachineApp/VendingMachineApp/SceneDelegate.swift:
```diff
@@ -12,8 +12,42 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+            if let vendingMachine = unarchive(with: data) {
+                appDelegate.vendingMachine = vendingMachine
+            }
+        }
```
AppDelegate와 차이점은 무엇인가요?  시점상으로 여기서만 하면 될까요? 
또는 iOS 버전과 관련없이 동작해야 하지 않을까요?

1Consumption replied:
맞습니다. 저는 iOS 13 버전만 고려 했었네요. 
그 하위 버전도 고려해주기 위해 available(iOS 13.0, *)을 사용하여 iOS 13 버전과 아닌 버전을 나눠서 실행하게 설정하였습니다. iOS 13에서는 UILifeCycle을 SceneDelegate에서 관리하고, 그 미만 버전에서는 AppDelegate에서 관리하기 때문입니다.   

* iOS 13 버전에서는
   ``` swift
   func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions:    [UIApplication.LaunchOptionsKey: Any]?)
   ```
   에서 언아카이빙을 해줬고
   ``` swift
   func sceneWillResignActive(_ scene: UIScene)
   ```
   에서 아카이빙을 해줬습니다.

* 그 미만 버전에서는 
   ``` swift
   func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?)
   ```
   에서 언아카이빙을 해줬고

   ``` swift
   func applicationWillResignActive(_ application: UIApplication) 
   ```
   에서 아카이빙을 해줬습니다.

11 버전 미만은 NSKeyedUnarchiver, NSKeyedarchiver를 지원하지 않아서 고려하지 않았습니다.

godrm commented on VendingMachineApp/VendingMachineApp/SceneDelegate.swift:
```diff
@@ -12,8 +12,42 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+            print(error)
+        }
+        return nil
+    }
```
SceneDelegate 역할과 책임은 무엇인가요? 
archive() unarchive() 구현을 Scene 마다 해야하는 이유는 무엇일까요?

1Consumption replied:
SceneDelegate에서는 UI life cycle을 관리합니다.

그런데 archive() unarchive() 메소드는 어플리케이션 전체에서 사용하는 메소드니까 AppDelegate에 구현하는 것이 더 맞다고 생각이 되네요. 또한 13 미만 버전의 경우는 Scene이 없으니까 더더욱 AppDelegate에 구현해야겠네요. 수정하였습니다.

godrm reviewed (commented):
enum 타입에 대한 아카이브 방법은 좋은 시도네요. 
그 외 내용에 대해 피드백을 남겼습니다.

1Consumption committed:
[기능 수정] archive, unarchive 메소드 위치 변경

이전의 경우 매 scene마다 동일한 역할을 하는 archive, unarchive 메소드가 있었지만
이를 AppDelegate로 옮겨 scene이 새로 생성되어도 해당 메소드를 구현하지 않게 함.

1Consumption committed:
[기능 수정] iOS 하위 버전 고려

iOS 13부터 SceneDelegate에서 Scene lifecycle에 맞게 행동을 취하도록 설정함.
13버전 미만은 AppDelegate에서 라이프 사이클에 맞게 행동을 취하도록 함.

1Consumption committed:
[기능 수정] Property 클래스 추가

기존 model 객체들은 각각 Property enum을 가지고 있었는데,
이를 Property 클래스에 String type으로 한꺼번에 관리함

1Consumption committed:
[기능 수정] if else 문 제거
